File: frontend/src/api/campaignLifecycle.ts
import { safeGet } from "@/api/api";

export type LifecycleStatusCount = {
  status: string;
  count: number;
  is_terminal: boolean;
};

export type LifecycleUnknownStatusCount = {
  status: string;
  count: number;
};

export type LifecycleEdge = {
  source: string;
  target: string;
};

export type CampaignLifecycleVisualization = {
  computed_at_utc: string;
  total_campaign_rows: number;
  status_counts: LifecycleStatusCount[];
  unknown_status_counts: LifecycleUnknownStatusCount[];
  edges: LifecycleEdge[];
  self_loop_statuses: string[];
  terminal_statuses: string[];
  bulk_transitions: string[];
  mermaid_state_diagram: string;
};

export function fetchCampaignLifecycle(): Promise<CampaignLifecycleVisualization> {
  return safeGet<CampaignLifecycleVisualization>("/campaigns/lifecycle");
}

File: backend/app/schemas/campaign_lifecycle.py
"""API schemas for read-only campaign lifecycle visualization."""

from __future__ import annotations

from pydantic import BaseModel, Field


class LifecycleStatusCount(BaseModel):
    status: str
    count: int = Field(..., ge=0)
    is_terminal: bool


class LifecycleUnknownStatusCount(BaseModel):
    status: str
    count: int = Field(..., ge=0)


class LifecycleEdge(BaseModel):
    source: str
    target: str


class LifecycleVisualizationResponse(BaseModel):
    computed_at_utc: str
    total_campaign_rows: int = Field(..., ge=0)
    status_counts: list[LifecycleStatusCount]
    unknown_status_counts: list[LifecycleUnknownStatusCount]
    edges: list[LifecycleEdge]
    self_loop_statuses: list[str]
    terminal_statuses: list[str]
    bulk_transitions: list[str]
    mermaid_state_diagram: str
